src/config_builder.py

```diff
@@ -112,24 +118,28 @@ def _ruler(self) -> dict:
 
     @property
     def _schema_config(self) -> dict:
-        return {
-            "configs": [
-                {
-                    "from": "2020-10-24",
-                    "index": {"period": "24h", "prefix": "index_"},
-                    "object_store": "filesystem",
-                    "schema": "v11",
-                    "store": "boltdb-shipper",
-                },
+        configs = [
+            {
+                "from": "2020-10-24",
+                "index": {"period": "24h", "prefix": "index_"},
+                "object_store": "filesystem",
+                "schema": "v11",
+                "store": "boltdb-shipper",
+            }
+        ]
+
+        if self.v12_migration_date:
```

Comment:
Rendering it conditionally is a nice touch!
Add a comment? Iiuc:
```suggestion
        # Only render the v12/tsdb section when we can reliably obtain a "from" date.
        # This is needed to avoid code ordering issues during upgrade, related to "can_connect".
        if self.v12_migration_date:
```